sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -2060,7 +2060,7 @@ case class TruncTimestamp(
   group = "datetime_funcs",
   since = "1.5.0")
 case class DateDiff(endDate: Expression, startDate: Expression)
-  extends BinaryExpression with ImplicitCastInputTypes with NullIntolerant {
+  extends BinaryExpression with ExpectsInputTypes with NullIntolerant {
```

Comment:
As this breaks Hive compatibility, I am limiting the scope of this PR to SubtractDates

```
org.apache.spark.sql.hive.execution.HiveCompatibilitySuite.udf9 | 0.5 sec | 1
org.apache.spark.sql.hive.execution.HiveQuerySuite.Cast Timestamp to Timestamp in UDF
```